MeshGeoToolsLib/GeoMapper.cpp

```diff
@@ -352,6 +356,342 @@ void GeoMapper::advancedMapOnMesh(
+{
+    std::vector<MathLib::Point3d> element_intersections;
+    for (std::size_t k(0); k < elem.getNumberOfEdges(); ++k)
+    {
+        auto const edge =
+            std::unique_ptr<MeshLib::Element const>(elem.getEdge(k));
+        GeoLib::LineSegment elem_segment{
+            new GeoLib::Point(*dynamic_cast<MathLib::Point3d*>(
+                const_cast<MeshLib::Node*>(edge->getNode(0))), 0),
+            new GeoLib::Point(*dynamic_cast<MathLib::Point3d*>(
+                const_cast<MeshLib::Node*>(edge->getNode(1))), 0),
+            false};
+        std::vector<MathLib::Point3d> const intersections(
+            GeoLib::lineSegmentIntersect2d(segment, elem_segment));
+        for (auto const& p : intersections)
+            element_intersections.emplace_back(p);
+    }
+    return element_intersections;
+}
+
+static std::vector<GeoLib::LineSegment> createSubSegmentsForElement(
+    std::vector<MathLib::Point3d> const& intersections,
+    MeshLib::Element const* const beg_elem,
+    MeshLib::Element const* const end_elem, MathLib::Point3d const& beg_pnt,
+    MathLib::Point3d const& end_pnt, MeshLib::Element const* const elem)
+{
+    std::vector<GeoLib::LineSegment> sub_segments;
+    if (intersections.size() > 2) {
+        std::stringstream out;
+        out << "element with id " << elem->getID() << " and seg "
+                  << " intersecting at more than two edges\n";
+        for (std::size_t k(0); k<intersections.size(); ++k)
+            out << k << " " << intersections[k]
+                      << "\n";
+        out << "Could not map segment on element. Aborting.\n";
+        ERR("%s", out.str().c_str());
+        std::abort();
+    }
+
+    if (intersections.size() == 1 && elem == beg_elem)
+    {
+        // The line segment intersects the element that contains the begin
+        // point of the line segment. Here the first sub line segment is
+        // added.
+        if (MathLib::sqrDist(beg_pnt, intersections[0]) >
+            std::numeric_limits<double>::epsilon())
+            sub_segments.emplace_back(GeoLib::LineSegment{
+                new GeoLib::Point{beg_pnt, 0},
+                new GeoLib::Point{intersections[0], 0}, true});
+    }
+
+    if (intersections.size() == 1 && elem == end_elem)
+    {
+        // The line segment intersects the element that contains the end
+        // point of the line segment. Here the last sub line segment is
+        // added.
+        if (MathLib::sqrDist(end_pnt, intersections[0]) >
+            std::numeric_limits<double>::epsilon())
+            sub_segments.emplace_back(
+                GeoLib::LineSegment{new GeoLib::Point{intersections[0], 0},
+                                    new GeoLib::Point{end_pnt, 0}, true});
+    }
+
+    if (intersections.size() == 1 && (elem != beg_elem && elem != end_elem))
+    {
+        // Since the line segment enters and leaves the element in the same
+        // point there isn't any need to insert a new sub line segment.
```

Comment:
Can that happen? If a line intersects an element in a point, i.e., in a corner of the element, it always intersects two edges, doesn't it? Or am I misunderstanding the entire matter?